cnxepub/formatters.py

```diff
@@ -426,7 +429,77 @@ def exercise_callback_factory(match, url_template,
     """Create a callback function to replace an exercise by fetching from
     a server."""
 
-    def _replace_exercises(elem):
+    def _annotate_exercise(elem, exercise, page_uuids):
+        """Annotate exercise based upon tag data"""
+        tags = exercise['items'][0].get('tags')
+        if not tags:
+            return
+
+        # Annotate exercise with required context, if it exists
+        modules, feature = [], ''
+        for tag in tags:
+            if 'context-cnxmod:' in tag:
+                modules.append(tag.split(':')[1])
+            if 'context-cnxfeature:' in tag:
+                # Note: Assuming this tag will never be duplicated with
+                # multiple values in the exercise data
+                feature = tag.split(':')[1]
+
+        # It's possible that the feature tag is not present, or present but
+        # not valid (e.g. value of "context-cnxfeature:").
+        if not feature:
+            return
+
+        # There may be multiple `context-cnxmod:{uuid}` tags, each of which
+        # could be the parent page for this exercise, a different page in this
+        # book, or even invalid altogether. We'll prefer the first, fallback
+        # to the second, and not annotate in the last case.
+
+        # Determine parent page for this exercise
+        parent_page_uuid = None
+        parent_page_elem = None
+        for parent in elem.iterancestors():
+            if parent.get('data-type') == 'page':
+                parent_page_elem = parent
+                parent_page_uuid = parent.get('id')
+                if parent_page_uuid.startswith('page_'):
+                    # Strip `page_` prefix from ID to get UUID
+                    parent_page_uuid = parent_page_uuid.split('page_')[1]
+                break
+
+        candidate_uuids = set(modules) & set(page_uuids)
+
+        # Check if the target feature ID is on the parent page for this
+        # exercise. If so, that takes priority over any context-cnxmod tag
+        # values, and should be picked even in cases where the parent page
+        # doesn't match one of the exercise tags. If the feature exists,
+        # we make sure the parent page UUID it's included in candidate_uuids
+        maybe_feature = parent_page_elem.find(
+            './/*[@id="{}"]'.format("auto_%s_%s" % (parent_page_uuid, feature))
+        )
+        if (maybe_feature is not None):
+            candidate_uuids.add(parent_page_uuid)
+
+        if len(candidate_uuids) == 0:
+            # No valid page UUIDs in exercise data
+            return
+
+        if parent_page_uuid in candidate_uuids:
+            target_module = parent_page_uuid
+        else:
+            # Use an UUID in the intersection of page UUIDs and tag UUIDs
+            # This is somewhat arbritrary, but if we hit this scenario with
+            # more than one UUID in the set things have gone pretty
+            # unexpectedly
```

Comment:
Just from reading this code it seems the formatter has the whole book inside self.root. If so, could you run one more check and throw if the `auto_{target_module}_{feature}` does not exist? That way it will be easier to debug than trying to find out why validation failed later on.

Reply:
Ah, so something like `find()` up to the body and search for the ID before inserting it? Should we just do this in general as a validation check before the code below that sets the context? (e.g. in case we select the current page UUID in spite of it not being there which I realize now the current code path would actually allow if it's in the tags)

Also, just as FYI: We can add a throw here, but one thing I realized is even if / where errors occur in these exercise processing threads they get silently tossed today and won't actually break `assemble` per [this recent bug](https://github.com/openstax/cnx/issues/1635) I created.

Reply:
I went ahead and [cleaned up](https://github.com/openstax/cnx-epub/pull/177/commits/b7974c51ae7916431789333b0f2b95882d85df26) the parent page check to discard that UUID if we find the feature doesn't exist there, and added a [general validation](https://github.com/openstax/cnx-epub/pull/177/commits/a5312c8271f85982f51a2e76587d7aa2e7554be2) of the final context determined. This should ensure that any invalid feature link will cause `assemble` to fail once [this bug](https://github.com/openstax/cnx/issues/1635) is fixed as well.